File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5Object.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.val;
import xyz.wagyourtail.commons.core.CollectionUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.*;

public class Json5Object extends StringData.OnlyParsed<Json5Object.ObjectContent> {


    public Json5Object(ObjectContent content) {
        super(content);
    }

    public static Json5Object parse(String raw) {
        StringCharReader charReader = new StringCharReader(raw);
        val content = contentBuilder(charReader);
        charReader.expectEOS();
        return new Json5Object(content);
    }

    public static Json5Object parse(CharReader<?> charReader) {
        return new Json5Object(contentBuilder(charReader));
    }

    public static ObjectContent contentBuilder(CharReader<?> reader) {
        List<Object> entries = new ArrayList<>();
        Map<String, Json5ObjectEntry> entryMap = new HashMap<>();
        reader.expect('{');
        while (!reader.exhausted()) {
            while (!reader.exhausted()) {
                val value = reader.parseOrNull(
                        Json5Whitespace::new,
                        Json5Comment::new
                );
                if (value == null) break;
                entries.add(value);
            }
            if (reader.peek() == '}') {
                break;
            }

            val entry = Json5ObjectEntry.parse(reader);
            entries.add(entry);
            entryMap.put(entry.getKey(), entry);

            while (!reader.exhausted()) {
                val value = reader.parseOrNull(
                        Json5Whitespace::new,
                        Json5Comment::new
                );
                if (value == null) break;
                entries.add(value);
            }
            if (reader.peek() == ',') {
                entries.add((char) reader.take());
            } else {
                break;
            }
        }
        reader.expect('}');
        return new ObjectContent(entries, entryMap);
    }

    @Getter
    @AllArgsConstructor
    public static class ObjectContent extends Data.Content {
        private final List<Object> entries;
        private final Map<String, Json5ObjectEntry> entryMap;

        public ObjectContent(List<Object> entries) {
            this.entries = entries;
            this.entryMap = new HashMap<>();
            for (Object entry : entries) {
                if (entry instanceof Json5ObjectEntry) {
                    Json5ObjectEntry objectEntry = (Json5ObjectEntry) entry;
                    entryMap.put(objectEntry.getKey(), objectEntry);
                }
            }
        }

        @Override
        public Iterable<Object> getEntries() {
            return CollectionUtils.concat(
                Collections.singleton('{'),
                entries,
                Collections.singleton('}')
            );
        }
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5Object.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.iterable

class Json5Object(content: ObjectContent) : StringData.OnlyParsed<Json5Object.ObjectContent>(content) {

    companion object : StringDataBuilder<Json5Object, ObjectContent> {

        override fun invoke(rawContent: CharReader<*>): Json5Object {
            return Json5Object(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): ObjectContent {
            val entries = mutableListOf<Any>()
            val entryMap = mutableMapOf<String, Json5ObjectEntry>()
            reader.expect('{')
            while (!reader.exhausted()) {
                while (!reader.exhausted()) {
                    val value = reader.parseOrNull(
                            Json5Whitespace::invoke,
                            Json5Comment::invoke
                    )
                    if (value == null) break
                    entries.add(value)
                }
                if (reader.peek() == '}') {
                    break
                }

                val entry = Json5ObjectEntry(reader)
                entries.add(entry)
                entryMap[entry.key] = entry

                while (!reader.exhausted()) {
                    val value = reader.parseOrNull(
                            Json5Whitespace::invoke,
                            Json5Comment::invoke
                    )
                    if (value == null) break
                    entries.add(value)
                }
                if (reader.peek() == ',') {
                    entries.add(reader.take()!!)
                } else {
                    break
                }
            }
            reader.expect('}')
            return ObjectContent(entries)
        }

    }

    class ObjectContent(
        val data: MutableList<Any>,
        val entryMap: MutableMap<String, Json5ObjectEntry>
    ) : Content() {

        constructor(data: MutableList<Any>): this(data, mutableMapOf()) {
            for (entry in data) {
                if (entry is Json5ObjectEntry) {
                    entryMap[entry.key] = entry
                }
            }
        }

        override val entries: Iterable<*>
            get() = iterable {
                yield('{')
                yieldAll(data)
                yield('}')
            }

    }

}